### PR TITLE
adding coverage overlay

### PR DESCRIPTION
**Pull request by bookman25**

### CODE REVIEW

**Comment by orta** [comment]
This looks 👍 - not sure what's up with those tests though

**Comment by bookman25** [comment]
All green now 😄

**Comment by orta** [comment]
amaze

File: src/Coverage/CoverageCodeLensProvider.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+export function registerCoverageCodeLens(jestExt: JestExt) {
+  return [
+    vscode.languages.registerCodeLensProvider(
+      { pattern: '**/*.{ts,tsx,js,jsx}' },
+      new CoverageCodeLensProvider(jestExt)
+    ),
+    vscode.commands.registerCommand(coverageCommand, () => {
+      vscode.window.showInformationMessage('You have unit tests!')
+    }),
```
Comment (seanpoulter) [inline src/Coverage/CoverageCodeLensProvider.ts]: Were we planning to use this command for anything? It looks like we could remove it now.

Comment (bookman25) [reply src/Coverage/CoverageCodeLensProvider.ts]: At one point I had a plan for it.  Not clue what it was now though :)  I think I was planning on maybe using it to toggle the coverage overlay (ended up doing that quick/dirty later though).

Comment (seanpoulter) [reply src/Coverage/CoverageCodeLensProvider.ts]: Oh, I know that feeling. I'll pop that out now.

**Comment by stephtr** [comment]
Is there any information or sample available on how to enable showing the coverage information?
I tried it for my own project, vscode-jest and create-react-app within vscode-jest, but without success.

**Comment by seanpoulter** [comment]
We don't have an example ready to go @stephtr but we'd appreciate a PR to improve the documentation or add to the example in `integrations/create-react-example`.
<br/>

You might find this repo I've been using to troubleshoot Issues helpful: [seanpoulter/vscode-jest-181](https://github.com/seanpoulter/vscode-jest-181). You'll see that Jest is configured to collect coverage in `jest.config.js`, and the setting is enabled in `.vscode/settings.json` to show the coverage when the test results come in.

**Comment by stephtr** [comment]
Thank you!

In order to enable the coverage informations when using create-react-app, one has to manually add `jest.pathToJest` to include `--coverage`.
Wouldn't it be more useful to add a new vscode setting (a la "collectCoverageByDefault") which automatically appends the `--coverage` flag? An advantage would be that one doesn't have to specify for Windows or Linux/Mac by including .cmd (`npm.cmd`) or not. That new setting would also work for other projects directly running jest (instead of `npm test`).

**Comment by seanpoulter** [comment]
> In order to enable the coverage informations when using create-react-app, one has to manually add jest.pathToJest to include --coverage.

You can also add it to the `scripts` in `package.json`. Since collecting coverage adds a performance hit  I've seen quite a few variations of `test` and `test-coverage` scripts.
<br/>

> Wouldn't it be more useful to add a new vscode setting (a la "collectCoverageByDefault") which automatically appends the `--coverage` flag?

That's an interesting idea. We welcome PRs if you're keen. For my day to day needs, I'd hold off using that setting until we could run tests related to the open files (#205) and improve the coverage overlay a bit (#177, local WIP).
<br/>

> An advantage would be that one doesn't have to specify for Windows or Linux/Mac by including .cmd (`npm.cmd`) or not.

Hrm. Does that mean a repo with `pathToJest` set wouldn't work across platforms? I think we'll need to fix this one at the root cause, and clean up the other issues related to running the right binaries: [#98](https://github.com/jest-community/vscode-jest/issues/98#issuecomment-353717341), #193.

**Comment by stephtr** [comment]
> You can also add it to the scripts in package.json. Since collecting coverage adds a performance hit I've seen quite a few variations of test and test-coverage scripts.

My thought was to have an option for enabling coverage overlay without having to edit scripts in package.json such that one could still manually run `npm test` without generating/displaying the generated coverage information (even though there is still the possibility to set `"jest": {"coverageReporters": []}` in package.json).

> That's an interesting idea. We welcome PRs if you're keen. For my day to day needs, I'd hold off using that setting until we could run tests related to the open files (#205) and improve the coverage overlay a bit (#177, local WIP).

The target audience in my mind were users with sparse experience who just run `create-react-app` (small projects). For them it would be more self-explaining and discoverable to toggle one VS Code setting than adding a script flag in package.json.

> Hrm. Does that mean a repo with pathToJest set wouldn't work across platforms? I think we'll need to fix this one at the root cause, and clean up the other issues related to running the right binaries: #98, #193.

vscode-jest sticks a `.cmd` to the end of a custom `jest.pathToJest` if not already present, which is not quite useful for `"jest.pathToJest": "npm test -- --coverage"` (or running with other flags). In order to prevent `Process failed: spawn npm ENOENT` on therefore has to write `"npm.cmd test -- --coverage"`, which isn't portable across different plattforms.

**Comment by seanpoulter** [comment]
I should've suggested this before -- let's move future discussion to the issues and give the poor CCed folks a rest.

**Pull request merged**